LibWeb: Move all CSS parsing out of StyleResolver

AtkinsSJ opened the pull request:
Another chonky PR, sorry! Removing `ValueListStyleValue` turned out to be more work than I expected. Still working on controlling my feature-creep tendencies too. :^)

My previous idea of `ValueListStyleValue` is a bit problematic: It means that parsing isn't complete when the Parser is done, and exposes internal Parser classes to the outside world. For correctness, we want to know if a property's value is invalid so that it can be rejected as soon as possible.

So, following the style of the `box-shadow` parsing code, this PR implements several `StyleValue` classes for specific style properties. In some cases, it is still useful to store a list of `StyleValue`s in a property, so this also implements `StyleValueList` for those cases. (eg, `border-width`, `border-radius`, `background` and `font-family`, which all take a variable-length list of values.)

Much of this is fairly mechanical: moving code from `StyleResolver` into methods in `Parser`. But in a couple of places I did get carried away: We now parse elliptical border radii, and the whole list of font names in `font` and `font-family`, and the `flex` parsing is more robust, among a few other small tweaks.

A larger difference is that the various `is_foo()` lambdas that used to be functions in `StyleResolver` no longer check for `calc()` and `var()`. This only affects properties that are handled specifically in `parse_css_value()`. (Which unfortunately does include `background` which is used by our `var()` test page, so I have switched that to `background-color`.) The reason is, that CSS variables can hold more than one value, eg `--awesome-variable: 12px 40px blue;`. This makes parsing them as part of a shorthand a lot more complicated than what we are currently doing, and I didn't want to get into that right now.

The first few commits here are about making it easier to deal with values that can be more than one thing. Likely, 

Since this implements a `BackgroundRepeatStyleValue`, the `background-repeat-x/y` pseudo-properties could be removed, but I have left them for now.

`ValueListStyleValue` was intended to be temporary in the first place, I just didn't realise it would be *this* temporary. But feels good to get rid of it!

TobyAsE commented on Base/res/html/misc/text-decoration.html:
```diff
@@ -4,15 +4,17 @@
 .overline { text-decoration: wavy blue overline; }
-.underline { text-decoration: double red underline; }
-.strikethrough { text-decoration: dotted green line-through; }
+.underline { text-decoration: red underline double; }
```
Isn't the old way also technically correct?

AtkinsSJ replied:
Yeah, this is just about testing that it still works in different orders.

TobyAsE commented on Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:
```diff
@@ -1503,7 +1503,17 @@ Optional<Length> Parser::parse_length(ParsingContext const& context, StyleCompon
+    // 2) It's a 0.
+    // We handle case 1 here. Case 2 is handled by NumericStyleValue pretending to be a LengthStyleValue if it is 0.
+
+    // FIXME: "auto" is also treated as a Length, and most of the time that is how it is used, but not always.
```
I feel your pain :^)

TobyAsE commented on Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:
```diff
@@ -1736,6 +1746,343 @@ RefPtr<StyleValue> Parser::parse_image_value(ParsingContext const& context, Styl
+    }
+}
+
+static inline bool is_background_image(StyleValue const& value)
```
Isn't this rather a `is_image()`?

AtkinsSJ replied:
I'm genuinely not sure. `none` might be a valid value for other places that accept an image, but it might not be. [The `background-image` spec](https://www.w3.org/TR/css-backgrounds-3/#background-image) says:
> Value: `<bg-image>#`
> ...
> Where
> `<bg-image> = <image> | none`

which implies that `<image>` and `none` are different things generally. I don't know. For now, I think it might be a trap to call it `is_image()` in case it's not true.

TobyAsE replied:
Better to stay with spec-language then.
I was just a bit confused because the function itself doesn't do anything that looks specifically 'background'-ish.

AtkinsSJ replied:
Yeah that's fair. And it started out as a lambda inside `parse_background_value()`, like the other ones, so I didn't think to change the name when I moved it.

There's sadly no way to check what values can be in CSS globally - it's all defined per property.

TobyAsE commented on Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:
```diff
@@ -1736,6 +1746,343 @@ RefPtr<StyleValue> Parser::parse_image_value(ParsingContext const& context, Styl
+    if (component_values.size() > 3)
+        return nullptr;
+
+    Optional<NonnullRefPtr<StyleValue>> maybe_border_width;
+    Optional<NonnullRefPtr<StyleValue>> maybe_border_color;
+    Optional<NonnullRefPtr<StyleValue>> maybe_border_style;
```
As above making those `RefPtr<>`s with `nullptr` representing the `sorry-ive-got-no-value` state seems to be the more Serenity-like approach.

TobyAsE commented on Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:
```diff
@@ -1777,6 +2124,594 @@ RefPtr<StyleValue> Parser::parse_box_shadow_value(ParsingContext const& context,
+    };
+
+    auto is_flex_basis = [](StyleValue const& value) -> bool {
+        if (value.is_length())
```
`auto` is covered by this?

AtkinsSJ replied:
It is currently, but that's not robust. I'll change it.

TobyAsE commented on Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:
```diff
@@ -1777,6 +2124,594 @@ RefPtr<StyleValue> Parser::parse_box_shadow_value(ParsingContext const& context,
+        if (!value)
+            return nullptr;
+
+        auto zero = NumericStyleValue::create(0);
+        auto one = NumericStyleValue::create(1);
+        auto auto_ident = IdentifierStyleValue::create(ValueID::Auto);
```
There are paths which don't need all of those. It's a bit sad to let the unused Values starve.

AtkinsSJ replied:
Sure. I realised while looking at this too, that the whole `is_initial()` case is unnecessary, if I update the values in Properties.json to have the correct initial values.

TobyAsE commented on Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:
```diff
@@ -1777,6 +2124,594 @@ RefPtr<StyleValue> Parser::parse_box_shadow_value(ParsingContext const& context,
+        return false;
+    };
+
+    Optional<NonnullRefPtr<StyleValue>> maybe_font_style;
+    Optional<NonnullRefPtr<StyleValue>> maybe_font_weight;
+    Optional<NonnullRefPtr<StyleValue>> maybe_font_size;
+    Optional<NonnullRefPtr<StyleValue>> maybe_line_height;
+    Optional<NonnullRefPtr<StyleValue>> maybe_font_families;
```
`RefPtr`

TobyAsE commented on Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:
```diff
@@ -1777,6 +2124,594 @@ RefPtr<StyleValue> Parser::parse_box_shadow_value(ParsingContext const& context,
+        if (part.is(Token::Type::String)) {
+            // `font-family: my cool "font";` is invalid.
+            if (!current_name_parts.is_empty())
+                return {};
```
`return nullptr` is what you do elsewhere. I think `RefPtr`s are defaulted to `nullptr`, so functionally there's no difference.
Same for all `return {}` in this function.

AtkinsSJ replied:
Good catch. The `parse_font_family_value()` changed quite a lot as I was working on it, and at one point it returned Optional. But since I'm using nullptr for everywhere else I'll change it to use that here too.

TobyAsE commented on Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:
```diff
@@ -1777,6 +2124,594 @@ RefPtr<StyleValue> Parser::parse_box_shadow_value(ParsingContext const& context,
+    if (component_values.size() > 3)
+        return nullptr;
+
+    Optional<NonnullRefPtr<StyleValue>> maybe_position;
+    Optional<NonnullRefPtr<StyleValue>> maybe_image;
+    Optional<NonnullRefPtr<StyleValue>> maybe_type;
```
`RefPtr`

AtkinsSJ replied:
:+1:

TobyAsE commented on Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:
```diff
@@ -1777,6 +2124,594 @@ RefPtr<StyleValue> Parser::parse_box_shadow_value(ParsingContext const& context,
+    if (component_values.size() > 3)
+        return nullptr;
+
+    Optional<NonnullRefPtr<StyleValue>> maybe_color;
+    Optional<NonnullRefPtr<StyleValue>> maybe_line;
+    Optional<NonnullRefPtr<StyleValue>> maybe_style;
```
`RefPtr`

TobyAsE reviewed (commented):
There are a few inconsistencies when doing `auto variable = parse_css_value()`. Sometimes `variable` is `maybe_something` and sometimes it's just `something`. Whatever you use, it should be consistent.
The `Optional<NonnullRefPtr<>>`s can be replaced with just `RefPtr`s, treating a `nullptr` as the `!Optional.has_value()`.

AtkinsSJ commented on Base/res/html/misc/border-radius.html:
```diff
@@ -105,12 +105,12 @@
             background-color: magenta;
-            border-top-left-radius: 10px;
+            border-radius: 15px;
         }
 
         .box-10 {
             background-color: magenta;
-            border-radius: 15px;
+            border-top-left-radius: 10px;
```
That's exactly what they are! I noticed they were the wrong way around, according to the text on the page, so I swapped them:
```html
    <em>All round 15px 5px thick</em>
    <div class="box box-9"></div>
    <br>
    <em>top-left 10px</em>
    <div class="box box-10"></div>
    <br>
```

TobyAsE replied:
Ah! Good catch!

AtkinsSJ commented on Base/res/html/misc/fonts.html:
```diff
@@ -5,9 +5,9 @@
             #monospace { font: 20px monospace; }
             #a { font: 12pt/14pt sans-serif; }
-            #b { font: 80% cursive; }
+            #b { font: bold 80% cursive; }
```
They're valid, and they did work without the change, but I wanted to test more combinations. That's what most of the .html file  changes are about. I probably should have spent more time originally on them to cover the various cases, but instead I'm modifying them as I go along. Again, I really need to get the official test suite to run so I don't have to worry so much. :^)

AtkinsSJ commented on Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:
```diff
@@ -1736,6 +1746,343 @@ RefPtr<StyleValue> Parser::parse_image_value(ParsingContext const& context, Styl
+
+RefPtr<StyleValue> Parser::parse_background_value(ParsingContext const& context, Vector<StyleComponentValueRule> const& component_values)
+{
+    Optional<NonnullRefPtr<StyleValue>> maybe_background_color;
+    Optional<NonnullRefPtr<StyleValue>> maybe_background_image;
+    Optional<NonnullRefPtr<StyleValue>> maybe_repeat_x;
+    Optional<NonnullRefPtr<StyleValue>> maybe_repeat_y;
```
Sure. I can't remember why I changed from RefPtr to this, but it's more complexity without any benefit. RefPtr is better.

AtkinsSJ commented on Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:
```diff
@@ -1736,6 +1746,343 @@ RefPtr<StyleValue> Parser::parse_image_value(ParsingContext const& context, Styl
+        auto& part = component_values[i];
+
+        // FIXME: Handle multiple backgrounds, by returning a List of BackgroundStyleValues.
+        if (part.is(Token::Type::Comma))
```
Good idea, it shouldn't be too bad.

AtkinsSJ commented on Base/res/html/misc/lists.html:
```diff
@@ -32,6 +32,18 @@ <h2>ul</h2>
     </ul>
 
+    <p>list-style: inside url(list-item.png)</p>
+    <ul style="list-style: inside disc url(list-item.png);">
```
:+1:

AtkinsSJ commented:
> There are a few inconsistencies when doing `auto variable = parse_css_value()`. Sometimes `variable` is `maybe_something` and sometimes it's just `something`. Whatever you use, it should be consistent.

Thanks for looking through it! I've corrected a few of these, I think it makes sense now?

> The `Optional<NonnullRefPtr<>>`s can be replaced with just `RefPtr`s, treating a `nullptr` as the `!Optional.has_value()`.

I think a lot of the weirdness is from me changing my mind about things repeatedly. At the time it made sense, but then I changed how things worked and didn't think to modify what I'd already done. Not using Optional also means avoiding allocating maybe-unused values in the `value_or(foo)` code from before, which is nice.
